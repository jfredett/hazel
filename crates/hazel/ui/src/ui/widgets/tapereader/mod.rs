use ratatui::{buffer::Buffer, layout::{Constraint, Layout, Rect}, style::{Style, Stylize}, text::Text, widgets::{Block, Row, StatefulWidget, Table, TableState, Widget}};
use hazel_representation::game::chess::state::tape_reader_state::TapeReaderState;

#[derive(Default)]
pub struct TapeReaderWidget {
    pub desired_position: usize
}

impl TapeReaderWidget {
    pub fn layout(&self) -> Layout {
        Layout::vertical([
            Constraint::Length(1), // header
            Constraint::Length(40), // code seciton
            Constraint::Min(1), // footer
        ])
    }

    pub fn advance(&mut self) {
        tracing::trace!(target="hazel::ui::events", "advancing to desired position: {:#04X} from {:#04X}", self.desired_position + 1, self.desired_position);
        self.desired_position += 1;
    }

    pub fn rewind(&mut self) {
        tracing::trace!(target="hazel::ui::events", "rewinding to desired position: {:#04X} from {:#04X}", self.desired_position.saturating_sub(1), self.desired_position);
        self.desired_position = self.desired_position.saturating_sub(1);
    }


    pub fn header_row(&self) -> Row {
        Row::new(vec!["Address", "Instruction", "Hash"])
            .style(Style::new().bold())
    }

    pub fn header(&self, state: &mut TapeReaderState) -> Text {
        let (pos, page, total_pages, len) = state.header();
        Text::from(
            format!("Tape: POS: {:#07X} ({}/{}), EOT: {:#06X}",
                pos, page, total_pages, len
            )
        )
    }

    pub fn table_state(&self, state: &mut TapeReaderState) -> TableState {
        TableState::default()
            .with_selected(state.position_in_page())
    }

    pub fn footer(&self) -> Text {
        Text::from("Footer here".to_string())
    }

    pub fn rows(&self, state: &mut TapeReaderState) -> Vec<Row> {
        let mut ret : Vec<Row> = state.context.clone().into_iter().enumerate().map(|(idx, e)| {
            // we have the alteration + context from `state` proper, we need to prepare the context
            // rows here, and add the header/footer rows (not sections) later.
            Row::new(vec![
                format!("{:#06X}", idx + state.offset()),
                e.to_string(),
                "Running Hash".to_string()
            ])
        }).collect();

        for addr in ret.len()..state.length {
            ret.push(Row::new(vec![format!("{:#06X}", state.offset() + addr)]));
        }

        ret
    }
}

impl StatefulWidget for &TapeReaderWidget {
    type State = TapeReaderState;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        // Columns widths are constrained in the same way as Layout...

        let chunks = self.layout().split(area);
        let header = chunks[0];
        let tape_area = chunks[1];
        let footer = chunks[2];

        let widths = [
            Constraint::Length(8),
            Constraint::Length(32),
            Constraint::Length(32)
        ];

        // TODO: This is probably not identity, but some function of height
        // state.set_page_size((code.height - 8) as usize);

        let table = Table::new(self.rows(state), widths)
            .column_spacing(1)
            .style(Style::new().white())
            .block(Block::bordered())
            .header(self.header_row())
            .row_highlight_style(Style::new().reversed())
            .column_highlight_style(Style::new().red())
            .cell_highlight_style(Style::new().blue())
            .highlight_symbol(">>");

        Widget::render(self.header(state), header, buf);
        StatefulWidget::render(&table, tape_area, buf, &mut self.table_state(state));
        Widget::render(self.footer(), footer, buf);
    }
}

#[cfg(test)]
mod tests {
    // use insta::assert_debug_snapshot;
    // use ratatui::{buffer::Buffer, style::Color};

    // use hazel_representation::{game::position::Position, notation::pgn::PGN};

    // use super::*;

    // // FIXME: familiar refactor
    // #[test]
    // fn renders_as_expected() {
    //     let rect = Rect::new(0, 0, 64, 32);
    //     let mut actual = Buffer::empty(rect);
    //     actual.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

    //     let pgn = PGN::load("tests/fixtures/no-variations-and-halts.pgn").unwrap();
    //     let mut position = Position::new(pgn.current_position());
    //     let tape = position.tape.read().unwrap();
    //     let mut fam = tape.conjure::<TapeReaderState>();

    //     let tapereader = TapeReaderWidget::default();

    //     tapereader.render(rect, &mut actual, fam.get_mut());

    //     assert_debug_snapshot!(actual);
    // }
}
