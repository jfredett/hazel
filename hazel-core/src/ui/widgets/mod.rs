// Sections are direct representations of chunks of the UI
pub mod tile;

pub mod engine_io_section;
pub mod game_section;

// other widgets are more general use oriented.
pub mod query;
pub mod placeholder;
pub mod tapereader;

pub mod input;
pub mod output;

pub mod board;
pub mod smallboard;
pub mod fen;

