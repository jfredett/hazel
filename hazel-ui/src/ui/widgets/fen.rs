use ratatui::{buffer::Buffer, prelude::*, text::Text, widgets::Widget};

use crate::notation::ben::BEN;


#[derive(Debug, Default)]
#[allow(clippy::upper_case_acronyms)]
pub struct FEN {
    fen: BEN,
    style: Style,
    alignment: Alignment
}

impl FEN {
    pub fn new(fen: BEN) -> Self {
        Self {
            fen,
            style: Style::default(),
            alignment: Alignment::Left
        }
    }

    pub fn set_style(mut self, style: Style) -> Self {
        self.style = style;
        self
    }

    pub fn alignment(mut self, alignment: Alignment) -> Self {
        self.alignment = alignment;
        self
    }

    #[inline(always)]
    pub fn center(self) -> Self {
        self.alignment(Alignment::Center)
    }
}

impl Widget for &FEN {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let fenstring = format!("{}", self.fen);
        Text::styled(fenstring, self.style).alignment(self.alignment).render(area, buf);
    }
}

#[cfg(test)]
mod tests {
    use crate::constants::START_POSITION_FEN;
    use crate::board::simple::PieceBoard;

    use super::*;

    #[test]
    fn renders_empty_fen_correctly() {
        let rect = Rect::new(0, 0, 64, 1);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let fen_widget = &FEN::new(BEN::empty());
        fen_widget.render(rect, &mut buffer);

        let mut expected = Buffer::with_lines(vec![
            "8/8/8/8/8/8/8/8 w KQkq - 0 1                                    "
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }

    #[test]
    fn render_startpos_fen_correctly() {
        let rect = Rect::new(0, 0, 64, 1);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let fen_widget = &FEN::new(BEN::new(START_POSITION_FEN));
        fen_widget.render(rect, &mut buffer);

        let mut expected = Buffer::with_lines(vec![
            "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1        "
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);

    }
}
