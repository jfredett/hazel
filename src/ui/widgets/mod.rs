pub mod hazel;
pub mod ply;
