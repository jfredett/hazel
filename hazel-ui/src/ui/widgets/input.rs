use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders};

#[derive(Default)]
pub struct Input {
    content: String
}

impl Input {
    pub fn push(&mut self, input: char) {
        self.content.push(input);
    }

    pub fn pop(&mut self) {
        self.content.pop();
    }

    pub fn content(&self) -> String {
        self.content.clone()
    }

    pub fn flush(&mut self) -> String {
        let content = self.content.clone();
        self.content.clear();
        content
    }
}

impl StatefulWidget for &Input {
    type State = String;

    fn render(self, area: Rect, buf: &mut Buffer, _state: &mut Self::State) {
        let block = Block::default()
            .borders(Borders::NONE)
            .title(format!("$> {}", self.content))
            .border_style(Style::default().fg(Color::White).bg(Color::Black));
        block.render(area, buf);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn renders_as_expected() {
        let rect = Rect::new(0, 0, 64, 3);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let input = &mut Input::default();
        input.push('a');
        input.push('b');
        input.push('c');
        input.render(rect, &mut buffer, &mut input.content());

        let mut expected = Buffer::with_lines(vec![
            "$> abc                                                          ",
            "                                                                ",
            "                                                                ",
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }

    #[test]
    fn flush_emptys_input_buffer() {
        let mut input = Input::default();
        input.push('a');
        input.push('b');
        input.push('c');
        assert_eq!(input.flush(), "abc");
        assert_eq!(input.content(), "");
    }
}
