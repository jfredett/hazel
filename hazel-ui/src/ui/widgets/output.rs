
use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders};

#[derive(Default)]
pub struct Output {
    buffer: Vec<String>
}

impl Output {
    pub fn push(&mut self, line: String) {
        self.buffer.push(line);
    }

    pub fn buffer(&self) -> Vec<String> {
        self.buffer.clone()
    }
}

fn wrap_string(s: &str, width: usize) -> Vec<String> {
    let mut lines = vec![];
    let mut working_copy = s.to_string();
    while working_copy.len() > width {
        let new_line = working_copy.split_off(width);
        lines.push(working_copy);
        working_copy = new_line;
    }
    lines.push(working_copy);
    lines
}

impl StatefulWidget for &Output {
    type State = Vec<String>;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let block = Block::default()
            .borders(Borders::ALL)
            .border_style(Style::default().fg(Color::White).bg(Color::Black));
        block.render(area, buf);

        // long lines are wrapped and indented by 2 spaces
        let output = state.clone();
        let mut adjusted = vec![];
        for line in output.iter() {
            if line.len() > area.width as usize - 2 {
                let new_lines = wrap_string(line, area.width as usize - 2);
                adjusted.extend(new_lines.into_iter());
            } else {
                adjusted.push(line.to_string());
            }
        }

        // should fill from the bottom to the top:
        let mut y = area.bottom() - 2;
        for line in adjusted.into_iter().rev() {
            buf.set_string(area.left() + 1, y, line, Style::default().fg(Color::White).bg(Color::Black));
            y -= 1;
        }
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn renders_as_expected() {
        let rect = Rect::new(0, 0, 64, 3);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let output = &mut Output::default();
        output.push("This is a test".to_string());
        output.render(rect, &mut buffer, &mut output.buffer().clone());

        let mut expected = Buffer::with_lines(vec![
            "┌──────────────────────────────────────────────────────────────┐",
            "│This is a test                                                │",
            "└──────────────────────────────────────────────────────────────┘",
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }

    #[test]
    fn renders_a_very_long_line_by_wrapping() {
        let rect = Rect::new(0, 0, 15, 6);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let output = &mut Output::default();
        output.push("This is a very long line that should be wrapped".to_string());
        output.render(rect, &mut buffer, &mut output.buffer().clone());

        let mut expected = Buffer::with_lines(vec![
        "┌─────────────┐",
        "│This is a ver│",
        "│y long line t│",
        "│hat should be│",
        "│ wrapped     │",
        "└─────────────┘",
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }
}







