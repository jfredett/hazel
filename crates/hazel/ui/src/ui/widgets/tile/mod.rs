use ratatui::layout::{Constraint, Direction, Layout, Rect};
use ratatui::widgets::Borders;
use ratatui::prelude::*;

use super::engine_io_section::EngineIOSection;
use super::fen::FEN;
use super::game_section::GameSectionLayout;
use super::placeholder::Placeholder;

use hazel::notation::pgn::PGN;

lazy_static! {
    static ref LAYOUT : Layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints(
            [
                Constraint::Length(19), // Game Section w/ Info and Board Sections
                Constraint::Length(1),  // Query Line, shown in the sketch w/ a FEN of the current position.
                Constraint::Length(12), // Engine IO Section
            ]
                .as_ref(),
        );
}

const WIDTH : u16 = 64;
const HEIGHT : u16 = 32;

#[derive(Default)]
pub struct Tile {
    engine_io_section: EngineIOSection,
    state: PGN,
}


impl Tile {
    pub fn new() -> Self {
        Self {
            engine_io_section: EngineIOSection::default(),
            // HACK: Obviously wrong, but pending loading up from a game/interacting with an engine
            // backend.
            state: PGN::load("tests/fixtures/no-variations-and-halts.pgn").unwrap(),
        }
    }

    pub fn handle_input(&mut self, input: char) {
        self.engine_io_section.handle_input(input);
    }

    pub fn handle_backspace(&mut self) {
        self.engine_io_section.handle_backspace();
    }

    pub fn handle_enter(&mut self) {
        self.engine_io_section.handle_enter();
    }

    pub fn set_state(&mut self, state: PGN) {
        self.state = state;
    }

    pub fn query_line(&self) -> FEN {
        FEN::new(self.state.current_position()).alignment(Alignment::Center)
    }
}

impl Widget for &Tile {
    fn render(self, area: Rect, buf: &mut Buffer) {
        // TODO: We always render at the same size, this will likely be wrong if the size is too
        // small, but I can add logic later.
        // Really I need an 'offset' Rect, the x,y components are added, the WIDTH/HEIGHT are min'd
        // against the provided area, if the area is too small, we render to an internal buffer and
        // cut the piece we out to match.
        let adjusted_area = Rect::new(area.x, area.y, WIDTH, HEIGHT);
        let chunks = LAYOUT.split(adjusted_area);

        let game_section = GameSectionLayout::new(self.state.clone());
        game_section.render(chunks[0], buf);
        Widget::render(
            &Placeholder::of_size(chunks[1].width, chunks[1].height).borders(Borders::LEFT | Borders::RIGHT),
            chunks[1],
            buf
        );

        self.engine_io_section.render(chunks[2], buf);

        self.query_line().render(chunks[1], buf);
        self.engine_io_section.render(chunks[2], buf);
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn renders_as_expected_in_larger_canvas() {
        let rect = Rect::new(0, 0, 65, 33);
        let mut actual = Buffer::empty(rect);
        actual.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let tile = Tile::new();
        tile.render(rect, &mut actual);

        assert_debug_snapshot!(actual);
    }
    #[test]
    fn renders_as_expected() {
        let rect = Rect::new(0, 0, 64, 32);
        let mut actual = Buffer::empty(rect);
        actual.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let tile = Tile::new();
        tile.render(rect, &mut actual);

        assert_debug_snapshot!(actual);
    }
}

