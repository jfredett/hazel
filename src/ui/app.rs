use std::collections::HashMap;
use std::fmt::Debug;

use ratatui::crossterm::event::{Event, KeyCode};
use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders};

use crate::ui::widgets::game_section::board_section::fenwidget::FENWidget;
use crate::ui::widgets::game_section::board_section::boardwidget::BoardWidget;

use tracing::{debug, instrument};

use crate::uci::UCIMessage;
use crate::ui::model::entry::{Entry, stockfish};
use crate::engine::Engine;
use crate::ui::model::pieceboard::PieceBoard;

use super::widgets::tile::Tile;

enum Mode {
    Insert,
    Command
}

pub struct Hazel {
    flags: HashMap<String, bool>,
    entry: Entry,
    // UI
    mode: Mode,
    tile: Tile,
}

impl Debug for Hazel {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Hazel")
            .field("flags", &self.flags)
            .finish()
    }
}

impl Hazel {
    #[instrument]
    pub fn new() -> Self {
        let mut s = Self {
            flags: HashMap::new(),
            mode: Mode::Command,
            entry: stockfish(),
            tile: Tile::new(),
        };

        s.entry.exec(UCIMessage::UCI);
        s.entry.exec(UCIMessage::IsReady);
        debug!("setting startpos");
        s.entry.exec(UCIMessage::Position("startpos moves d2d4".to_string(), vec![]));
        debug!("setting startpos done");

        // s.entry.boardstate.set_startpos();
        s.entry.boardstate = PieceBoard::from_fen("rnbqkbnr/pppppppp/8/8/3P4/8/PPP1PPPP/RNBQKBNR");

        return s;
    }

    #[instrument]
    pub fn handle_events(&mut self, event: Event) {
        if let Event::Key(key) = event {
            match self.mode {
                // Probably insert mode is just handled by the tile wholesale?
                Mode::Insert => {
                    match key.code {
                        KeyCode::Esc => {
                            self.mode = Mode::Command;
                        },
                        KeyCode::Char(c) => {
                            self.tile.handle_input(c);
                        },
                        KeyCode::Backspace => {
                            self.tile.handle_backspace();
                        },
                        KeyCode::Enter => {
                            self.tile.handle_enter();
                        },
                        _ => {
                        }
                    }
                },
                // Command mode will eventually select the tile you want/start new tiles, etc.
                Mode::Command => {
                    match key.code {
                        KeyCode::Char('i') => {
                            self.mode = Mode::Insert;
                        },
                        KeyCode::Char('q') => {
                            self.set_flag("exit", true);
                        },
                        _ => {
                        }
                    }
                }
            }
        }
    }

    #[instrument]
    pub fn set_flag(&mut self, flag: &str, value: bool) {
        self.flags.insert(flag.to_string(), value);
    }

    #[instrument]
    pub fn check_flag(&self, flag: &str) -> bool {
        match self.flags.get(flag) {
            Some(value) => *value,
            None => false
        }
    }

    #[instrument]
    pub fn fen_widget(&self) -> FENWidget {
        FENWidget::from(&self.entry)
    }

    #[instrument]
    pub fn board_widget(&self) -> BoardWidget {
        BoardWidget::from(&self.entry.boardstate)
    }

    #[instrument]
    pub fn input_widget(&self) -> Block {
        Block::default()
            .title("Input")
            .borders(Borders::ALL)
    }


    #[instrument]
    pub fn render(&mut self, frame: &mut Frame) {
        self.tile.set_state(self.entry.boardstate.clone());
        frame.render_widget(&self.tile, Rect::new(0,0,64,32));
    }
}

#[cfg(test)]
mod tests {
    use std::process::Termination;

    use backend::TestBackend;

    use super::*;

    #[test]
    fn placeholder() {
        let mut hazel = Hazel::new();

        let mut t = Terminal::new(TestBackend::new(64, 32)).unwrap();
        let _ = t.draw(|f| {
            hazel.render(f);
        });

        let mut expected = Buffer::with_lines(vec![
            "           Placeholder           ┌─────────────────────────────┐",
            "           Placeholder           │         Placeholder         │",
            "           Placeholder           │         Placeholder         │",
            "┌───────────────┐┌──────────────┐│         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder │└─────────────────────────────┘",
            "└───────────────┘└──────────────┘          Placeholder          ",
            "│                          Placeholder                         │",
            "┌──────────────────────────────────────────────────────────────┐",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "└──────────────────────────────────────────────────────────────┘",
            "$>                                                              ",
        ]);

        // Ignore style differences for now
        let mut actual = t.backend().buffer().clone();
        actual.set_style(Rect::new(0, 0, 64, 32), Style::default().fg(Color::White).bg(Color::Black));
        expected.set_style(Rect::new(0, 0, 64, 32), Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(actual, expected);
    }
}
