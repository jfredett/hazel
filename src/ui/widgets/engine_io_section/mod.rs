pub mod outputwidget;

use ratatui::layout::Direction;
use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders};
use crate::ui::app::Hazel;

// TODO: minimize the state by making EngineIOSection not need all of Hazel, instead it'll
// be it's own state management ViewModel and Hazel will construct it from itself.

pub struct EngineIOSection {
    output: Vec<String>,
    input: String,
}


struct Input {
}

struct Output {
}

impl Default for Input {
    fn default() -> Self {
        Self {}
    }
}

impl Default for Output {
    fn default() -> Self {
        Self {}
    }
}

impl StatefulWidget for &Input {
    type State = String;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let block = Block::default()
            .borders(Borders::NONE)
            .title(format!("$> {}", state))
            .border_style(Style::default().fg(Color::White).bg(Color::Black));
        block.render(area, buf);
    }
}

impl StatefulWidget for &Output {
    type State = Vec<String>;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let block = Block::default()
            .borders(Borders::ALL)
            .border_style(Style::default().fg(Color::White).bg(Color::Black));
        block.render(area, buf);

        // long lines are wrapped and indented by 2 spaces
        let output = state.clone();
        let mut adjusted = vec![];
        for line in output.iter() {
            if line.len() > area.width as usize - 2 {
                let mut working_copy = line.clone();
                let mut new_lines = vec![working_copy.split_off(area.width as usize - 2)];

                while working_copy.len() > area.width as usize - 2 {
                    let new_line = format!("  {}", working_copy.split_off(area.width as usize - 4));
                    new_lines.push(new_line);
                }
                new_lines.push(working_copy);
                adjusted.extend(new_lines.into_iter().rev());
            } else {
                adjusted.push(line.to_string());
            }
        }

        // should fill from the bottom to the top:
        let mut y = area.bottom() - 2;
        for line in adjusted.into_iter().rev() {
            buf.set_string(area.left() + 1, y, line, Style::default().fg(Color::White).bg(Color::Black));
            y -= 1;
        }
    }
}


lazy_static! {
    static ref LAYOUT : Layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints(
            [
                Constraint::Fill(1),
                Constraint::Max(1),
            ]
                .as_ref(),
        );
}

impl EngineIOSection {
    pub fn new() -> Self {
        Self {
            output: vec![],
            input: String::new(),
        }
    }

    pub fn push(&mut self, line: String) {
        self.output.push(line);
    }
}

impl StatefulWidget for &EngineIOSection {
    type State = Hazel;
    fn render(self, area: Rect, buf: &mut Buffer, _state: &mut Self::State) {
        let chunks = LAYOUT.split(area);


        let output_widget = Output::default();
        output_widget.render(chunks[0], buf, &mut self.output.clone());

        let input_widget = Input::default();
        input_widget.render(chunks[1], buf, &mut self.input.clone());
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn placeholder() {
        let rect = Rect::new(0, 0, 64, 17);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let engine_io_section = &mut EngineIOSection::new();

        // Mock out the output from the 'engine'
        engine_io_section.push("Stockfish 16.1 by the Stockfish developers (see AUTHORS file)".to_string());
        engine_io_section.push("> isready".to_string());
        engine_io_section.push("readyok".to_string());
        engine_io_section.push("> position startpos moves".to_string());
        engine_io_section.push("> go depth 1".to_string());
        engine_io_section.push("info string NNUE evaluation using nn-baff1ede1f90.nnue".to_string());
        engine_io_section.push("info string NNUE evaluation using nn-b1a57edbea57.nnue".to_string());
        engine_io_section.push("info depth 1 seldepth 2 multipv 1 score cp 0 nodes 20 nps 1000 hashfull 0 tbhits 0 time 20 pv d2d4".to_string());
        engine_io_section.push("bestmove d2d4".to_string());

        engine_io_section.render(rect, &mut buffer, &mut Hazel::new());

        let mut expected = Buffer::with_lines(vec![
            "┌──────────────────────────────────────────────────────────────┐",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│Stockfish 16.1 by the Stockfish developers (see AUTHORS file) │",
            "│> isready                                                     │", // prefixes inputs by >
            "│readyok                                                       │",
            "│> position startpos moves                                     │",
            "│> go depth 1                                                  │",
            "│info string NNUE evaluation using nn-baff1ede1f90.nnue        │",
            "│info string NNUE evaluation using nn-b1a57edbea57.nnue        │",
            "│info depth 1 seldepth 2 multipv 1 score cp 0 nodes 20 nps 1000│", // wraps long lines?
            "│ hashfull 0 tbhits 0 time 20 pv d2d4                          │",
            "│bestmove d2d4                                                 │",
            "└──────────────────────────────────────────────────────────────┘",
            "$>                                                              ",
        ]);

        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));
        assert_eq!(buffer, expected);
    }
}


