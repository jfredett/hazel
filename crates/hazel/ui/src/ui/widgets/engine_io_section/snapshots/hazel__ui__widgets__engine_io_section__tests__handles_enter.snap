---
source: src/ui/widgets/engine_io_section/mod.rs
expression: buffer
---
Buffer {
    area: Rect { x: 0, y: 0, width: 64, height: 17 },
    content: [
        "┌──────────────────────────────────────────────────────────────┐",
        "│                                                              │",
        "│                                                              │",
        "│                                                              │",
        "│                                                              │",
        "│                                                              │",
        "│                                                              │",
        "│                                                              │",
        "│                                                              │",
        "│                                                              │",
        "│                                                              │",
        "│                                                              │",
        "│                                                              │",
        "│                                                              │",
        "│> abcdefg                                                     │",
        "└──────────────────────────────────────────────────────────────┘",
        "$>                                                              ",
    ],
    styles: [
        x: 0, y: 0, fg: White, bg: Black, underline: Reset, modifier: NONE,
    ]
}
